include/boost/compute/algorithm/detail/find_extrema_reduce.hpp

```diff
@@ -0,0 +1,245 @@
+    const size_t local_mem_size = device.get_info<CL_DEVICE_LOCAL_MEM_SIZE>();
+
+    size_t work_group_size = (std::min)(max_work_group_size, preferred_work_group_size);
+
+    // local memory size needed to perform parallel reduction
+    size_t required_local_mem_size = 0;
+    // indices size
+    required_local_mem_size += sizeof(uint_) * work_group_size;
+    // values size
+    required_local_mem_size += sizeof(input_type) * work_group_size;
+
+    // at least 4 work groups per compute unit otherwise reduction
+    // would be highly inefficient
+    return ((required_local_mem_size * 4) <= local_mem_size);
+}
+
+template<class InputIterator, class ResultIterator>
+inline size_t find_extrema_reduce(InputIterator first,
+                                  size_t count,
+                                  ResultIterator result,
+                                  vector<uint_>::iterator result_idx,
+                                  size_t work_groups_no,
+                                  size_t work_group_size,
+                                  char sign,
+                                  command_queue &queue)
+{
+    typedef typename std::iterator_traits<InputIterator>::value_type input_type;
+
+    const context &context = queue.get_context();
+
+    meta_kernel k("find_extrema_reduce");
+
+    k.add_extension_pragma("cl_amd_printf");
+
+    size_t count_arg = k.add_arg<uint_>("count");
+    size_t output_arg = k.add_arg<input_type *>(memory_object::global_memory, "output");
+    size_t output_idx_arg = k.add_arg<uint_ *>(memory_object::global_memory, "output_idx");
+    size_t block_arg = k.add_arg<input_type *>(memory_object::local_memory, "block");
+    size_t block_idx_arg = k.add_arg<uint_ *>(memory_object::local_memory, "block_idx");
+
+    k <<
+        // Work item global id
+        k.decl<const uint_>("gid") << " = get_global_id(0);\n" <<
+        //
+        "if(gid >= count) {\n return;\n }\n" <<
+
+        // Index of element that will be read from input buffer
+        k.decl<uint_>("idx") << " = gid;\n" <<
+
+        k.decl<input_type>("acc") << ";\n" <<
+        // Index of currently best element
+        k.decl<uint_>("acc_idx") << " = idx;\n" <<
+
+        "if( idx < count ){\n" <<
+            // Init accumulator with first[get_global_id(0)]
+            "acc = " << first[k.var<uint_>("idx")] << ";\n" <<
+            "idx += get_global_size(0);\n" <<
+        "}\n" <<
+
+        k.decl<bool>("compare_result") << ";\n" <<
+        "while( idx < count ){\n" <<
+            // Next element
+            k.decl<input_type>("next") << " = " << first[k.var<uint_>("idx")] << ";\n" <<
+            // Comparison between currently best element (acc) and next element
+            "compare_result = acc " << sign << " next;\n" <<
+            "acc = compare_result ? acc : next;\n" <<
+            "acc_idx = compare_result ? acc_idx : idx;\n" <<
+            "idx += get_global_size(0);\n" <<
```

Comment:
I am not entirely sure. My guess would be fewer cache misses. If the data is not aligned to the cache boundary, each read will need to get two cache lines per global stride. If using a local stride, the remaining data from the second cache line will be useful for the second iteration.